packages/react-admin/src/mui/list/FilterForm.js

```diff
@@ -172,15 +192,40 @@ export const mergeInitialValuesWithDefaultValues = ({
     },
 });
 
+/*
+ *
+ * 1. Limits the request rate of requests to the server
+ * 2. Allows the change to be properly propogated to the redux state
+ */
+const filterFormProps = ({ resource, debounce: debounceDelay }) => {
+    const onChange = debounce(
+        (values, dispatch, props) =>
+            props && props.form && dispatch && dispatch(submit(props.form)),
+        debounceDelay
+    );
+    return {
+        // We can't use the same form key for all resources, because the filter is cloned before rendering
+        // when the actual render will happen, the previous filter form will unmount and destroy the redux state for it's form
+        // So for a short period of time there are at most two FilterForm state's in the reduxform state.
+        form: `${resource}FilterForm`,
+        onChange,
+        cancelOnChangeDebounce: () => onChange.cancel(),
+    };
+};
+
+const combineWithProps = props => ({
+    ...mergeInitialValuesWithDefaultValues(props),
+    ...filterFormProps(props),
+});
+
 const enhance = compose(
     withStyles(styles),
     translate,
-    withProps(mergeInitialValuesWithDefaultValues),
+    withProps(combineWithProps),
```

Comment:
I don't understand why you add the `onChange` and `cancelOnChangeDebounce` methods as props rather than just methods of the component class. It obfuscates the code IMO.

Reply:
onChange is a prop of the `reduxForm` hoc, so either the form should be moved in it's own component (something like `FilterInnerForm`) decorated by `reduxForm` and pass the `onChange` prop to it using component method class. Or this solution needs to be taken. Simply putting the debounce on the `reduxForm` decorator, like:
```
reduxForm({
   onChange: debounce(...
)}
```
does not work, because this would hide the `cancel` function of debounce. 
So I opted for this solution, but have no problem with externalizing the component. Let me know.